src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/LspFileBasedProgramWorkspace.cs

```diff
@@ -0,0 +1,323 @@
+
+            var tree = CSharpSyntaxTree.ParseText(text: documentText, options: csharpOptions, path: documentFilePath);
+            var root = tree.GetRoot();
+            return
+                //root.DescendantNodes(descendIntoChildren: node => node.IsKind(SyntaxKind.CompilationUnit)).OfType<GlobalStatementSyntax>().Any() ||
+                root.GetLeadingTrivia().FirstOrDefault().IsKind(SyntaxKind.ShebangDirectiveTrivia);
+        }
+    }
+
+    private async ValueTask LoadOrReloadProjectsAsync(ImmutableSegmentedList<(Uri uri, SourceText documentText, string languageId)> list, CancellationToken token)
+    {
+        foreach (var pair in list)
+        {
+            await LoadOrReloadProjectAsync(pair, logger: null, token);
+        }
+    }
+
+    /// <summary>
+    /// Takes in a file URI and text and creates a misc project and document for the file.
+    /// 
+    /// Calls to this method and <see cref="TryRemoveMiscellaneousDocument(Uri, bool)"/> are made
+    /// from LSP text sync request handling which do not run concurrently.
+    /// </summary>
+    public async Task<Document?> LoadOrReloadProjectAsync((Uri uri, SourceText documentText, string languageId) pair, ILspLogger? logger, CancellationToken cancellationToken)
+    {
+        var (uri, documentText, languageId) = pair;
+        var documentFilePath = ProtocolConversions.GetDocumentFilePathFromUri(uri);
+
+        var existingProjects = _loadedProjects.GetOrAdd(documentFilePath, static _ => new List<LoadedProject>());
+
+        var languageInfoProvider = _lspServices.GetRequiredService<ILanguageInfoProvider>();
+        if (!languageInfoProvider.TryGetLanguageInformation(uri, languageId, out var languageInformation))
+        {
+            // Only log here since throwing here could take down the LSP server.
+            logger?.LogError($"Could not find language information for {uri} with absolute path {documentFilePath}");
+            return null;
+        }
+
+        await using var buildHostProcessManager = new BuildHostProcessManager(globalMSBuildProperties: null, binaryLogPath: null, loggerFactory: _loggerFactory);
+        var (buildHost, actualBuildHostKind) = await buildHostProcessManager.GetBuildHostWithFallbackAsync(
+            buildHostKind: BuildHostProcessManager.BuildHostProcessKind.NetCore,
+            // nb: file-based project is also source code
+            projectOrSolutionFilePath: documentFilePath,
+            cancellationToken);
+
+        var virtualProjectContent = MakeVirtualProjectContent(documentFilePath, documentText);
+        var loadedFile = await buildHost.LoadProjectAsync(Path.ChangeExtension(documentFilePath, "csproj"), virtualProjectContent, languageInformation.LanguageName, cancellationToken);
+        var loadedProjectInfos = await loadedFile.GetProjectFileInfosAsync(cancellationToken);
+        var loadedProjectInfo = loadedProjectInfos.First();
+        if (existingProjects is [var existingProject, ..])
```

Comment:
this is not safe.  this method may be called cconcurrently (you expose it outwards).  so it may be enumerated on one thread, while mutated on another.  that can crash depending on what each thread sees as mutation and reading is happening.